File: chat-service/src/main/java/com/oognuyh/chatservice/config/FeignConfig.java
package com.oognuyh.chatservice.config;

import org.springframework.context.annotation.Bean;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ResponseStatusException;

import feign.RequestInterceptor;
import feign.RequestTemplate;
import feign.Response;
import feign.codec.ErrorDecoder;

@Component
public class FeignConfig {
    
    @Bean
    public RequestInterceptor requestInterceptor() {
        return new RequestInterceptor() {

            @Override
            public void apply(RequestTemplate template) {
                Jwt jwt = (Jwt) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
                
                template.header(HttpHeaders.AUTHORIZATION, String.format("Bearer %s", jwt.getTokenValue()));
            }
        };
    }

    @Bean
    public ErrorDecoder errorDecoder() {
        return new ErrorDecoder() {

            @Override
            public Exception decode(String methodKey, Response response) {
                
                switch (response.status()) {
                    case 400:
                        return new ResponseStatusException(HttpStatus.BAD_REQUEST, response.reason());
                    case 404:
                        return new ResponseStatusException(HttpStatus.NOT_FOUND, response.reason());
                    default:
                        return new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, response.reason());
                }
            }
        };
    }
}

File: notification-service/src/main/kotlin/com/oognuyh/notificationservice/config/FeignConfig.kt
package com.oognuyh.notificationservice.config

import feign.codec.ErrorDecoder
import org.springframework.cloud.openfeign.EnableFeignClients
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.HttpStatus
import org.springframework.web.server.ResponseStatusException

@Configuration
@EnableFeignClients(basePackages = ["com.oognuyh.notificationservice.repository"])
class FeignConfig {

    @Bean
    fun errorDecoder(): ErrorDecoder {
        return ErrorDecoder { _, response ->
            when (response?.status()) {
                400 -> ResponseStatusException(HttpStatus.BAD_REQUEST)
                401 -> ResponseStatusException(HttpStatus.UNAUTHORIZED)
                403 -> ResponseStatusException(HttpStatus.FORBIDDEN)
                404 -> ResponseStatusException(HttpStatus.NOT_FOUND)
                else -> ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR)
            }
        }
    }
}